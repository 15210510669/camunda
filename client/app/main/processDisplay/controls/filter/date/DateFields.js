import React from 'react';
import {DateRange} from './DateRange';
import {DateInput} from './DateInput';
import classwrap from 'classwrap';

const jsx = React.createElement;
const $document = document; // for mocking
const $setTimeout = setTimeout; // for mocking

export class DateFields extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      popupOpen: false,
      currentlySelectedField: null,
      minDate: null,
      maxDate: null
    };
  }

  componentDidUpdate() {
    const {popupOpen} = this.state;

    if (popupOpen) {
      $document.addEventListener('click', this.hidePopup);
    } else {
      $document.removeEventListener('click', this.hidePopup);
    }
  }

  componentWillUnmount() {
    $document.removeEventListener('click', this.hidePopup);
  }

  render() {
    return <center className="date-fields">
      <div className="input-group input-daterange">
        <DateInput
            className={
              classwrap([
                'form-control',
                'start',
                {
                  'date-field-highlight': this.isFieldSelected('startDate')
                }
              ])
            }
            format={this.props.format}
            onDateChange={this.setStartDate}
            onClick={this.toggleDateRangeForStart}
            date={this.props.startDate} />
        <span className="input-group-addon">to</span>
        <DateInput
            className={
              classwrap([
                'form-control',
                'end',
                {
                  'date-field-highlight': this.isFieldSelected('endDate')
                }
              ])
            }
            reference={this.saveEndDateField}
            format={this.props.format}
            onDateChange={this.setEndDate}
            onClick={this.toggleDateRangeForEnd}
            date={this.props.endDate} />
      </div>
      <div onClick={this.stopClosingPopup} className={
        classwrap([
          'date-fields-range',
          {
            hidden: !this.state.popupOpen,
            'left-side': this.isFieldSelected('startDate'),
            'right-side': this.isFieldSelected('endDate')
          }
        ])
      }>
        <DateRange
            format={this.props.format}
            onDateChange={this.onDateRangeChange}
            startDate={this.props.startDate}
            endDate={this.props.endDate} />
      </div>
    </center>;
  }

  saveEndDateField = input => this.endDateField = input;

  onDateRangeChange = date => {
    this.props.onDateChange(this.state.currentlySelectedField, date);

    if (this.isFieldSelected('startDate')) {
      this.setState({
        currentlySelectedField: 'endDate',
      });
    } else {
      $setTimeout(this.hidePopup, 350);
    }
  }

  stopClosingPopup = ({nativeEvent: event}) =>  {
    // https://stackoverflow.com/questions/24415631/reactjs-syntheticevent-stoppropagation-only-works-with-react-events
    event.stopImmediatePropagation();
  }

  hidePopup = () => {
    this.setState({
      popupOpen: false,
      currentlySelectedField: null,
      minDate: null,
      maxDate: null
    });
  }

  isFieldSelected(field) {
    return this.state.currentlySelectedField === field;
  }

  setStartDate = date => this.props.onDateChange('startDate', date);
  setEndDate = date => this.props.onDateChange('endDate', date);

  toggleDateRangeForStart = event => this.toggleDateRangePopup(event, 'startDate');
  toggleDateRangeForEnd = event => this.toggleDateRangePopup(event, 'endDate');

  toggleDateRangePopup = (event, field) => {
    this.stopClosingPopup(event);

    if (this.state.popupOpen) {
      return this.setState({
        currentlySelectedField: field
      });
    }

    this.setState({
      popupOpen: true,
      currentlySelectedField: field
    });
  }
}
