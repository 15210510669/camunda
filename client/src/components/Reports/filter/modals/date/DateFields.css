.DateFields {
  position: relative;
}

.DateFields__inputContainer {
  display: flex;
}

.DateFields__range {
  position: absolute;
  left: 0;
  top: 50px;
  box-shadow: 0 5px 15px rgba(0, 0, 0, 0.5);
  border: 1px solid rgba(0, 0, 0, 0.2);
  background-color: white;
  padding: 5px 0;
}

.DateFields__range::before {
  content: ' ';
  display: block;
  width: 20px;
  height: 20px;
  position: absolute;
  transform: rotate(45deg);
  box-shadow: 0 5px 10px rgba(0, 0, 0, 0.5);
  border: 1px solid rgba(0, 0, 0, 0.2);
  background-color: white;
  top: -10px
}

.DateFields__range::after {
  content: ' ';
  display: block;
  width: 50px;
  height: 22px;
  position: absolute;
  background-color: white;
  top: 0;
}

.DateFields__range--left::before {
  left: 20px;
}
.DateFields__range--left::after {
  left: 0;
}

.DateFields__range--right::before {
  right: 20px;
}
.DateFields__range--right::after {
  right: 0;
}
