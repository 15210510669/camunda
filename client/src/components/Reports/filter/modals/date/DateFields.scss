.DateFields {
  position: relative;
  margin-bottom: 12px;
}

.DateFields__inputContainer {
  display: flex;
  justify-content: center;
}

.DateFields__range {
  position: absolute;
  left: 0;
  top: 50px;
  box-shadow: 0 5px 15px rgba(0, 0, 0, 0.5);
  border: 1px solid rgba(0, 0, 0, 0.2);
  background-color: white;
  padding: 5px 0;
  z-index: 9999;
}

.DateFields__range::before {
  content: ' ';
  display: block;
  width: 20px;
  height: 20px;
  position: absolute;
  transform: rotate(45deg);
  box-shadow: 0 5px 10px rgba(0, 0, 0, 0.5);
  border: 1px solid rgba(0, 0, 0, 0.2);
  background-color: white;
  top: -10px;
}

.DateFields__range::after {
  content: ' ';
  display: block;
  width: 50px;
  height: 22px;
  position: absolute;
  background-color: white;
  top: 0;
}

.DateFields__range--left::before {
  left: 20px;
}
.DateFields__range--left::after {
  left: 0;
}

.DateFields__range--right::before {
  right: 20px;
}
.DateFields__range--right::after {
  right: 0;
}

.DateFields__divider {
  color: #626262;
  text-align: center;
  font-size: 0.875em;
  display: flex;
  justify-content: center;
  flex-direction: column;
  width: 40px;
  background-color: #fafafa;
  border-top: 1px solid lightgray;
  border-left: none;
  border-right: none;
  border-bottom: 1px solid lightgray;
}
