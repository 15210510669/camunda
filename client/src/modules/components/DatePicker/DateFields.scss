@import 'shared-styles/shared-styles.scss';

.DateFields {
  position: relative;
  margin-bottom: 15px;

  .inputContainer {
    display: flex;
    align-items: center;
    flex-wrap: nowrap;

    .PickerDateInput:first-child {
      margin-right: 15px;
      &:after {
        right: -10px;
        position: absolute;
        top: 50%;
        transform: translateY(-50%);
        content: '-';
      }
    }
  }

  .dateRangeContainer {
    position: absolute;
    width: max-content;
    left: 0;
    top: 50px;
    box-shadow: 0 5px 15px rgba(0, 0, 0, 0.5);
    border: 1px solid rgba(0, 0, 0, 0.2);
    background-color: white;
    padding: 5px 0;
    z-index: var(--z-popup);

    &.dateRangeContainerLeft {
      &::before {
        left: 20px;
      }

      &::after {
        left: 0;
      }
    }

    &.dateRangeContainerRight {
      &::before {
        left: 160px;
      }
      &::after {
        left: 140px;
      }
    }

    &::before {
      content: ' ';
      display: block;
      width: 20px;
      height: 20px;
      position: absolute;
      transform: rotate(45deg);
      border: 1px solid rgba(0, 0, 0, 0.2);
      background-color: white;
      top: -10px;
    }

    &::after {
      content: ' ';
      display: block;
      width: 50px;
      height: 22px;
      position: absolute;
      background-color: white;
      top: 0;
    }
  }
}
