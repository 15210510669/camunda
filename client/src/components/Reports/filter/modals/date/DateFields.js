import React from 'react';
import DateRange from './DateRange';
import DateInput from './DateInput';

import './DateFields.css';

export default class DateFields extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      popupOpen: false,
      currentlySelectedField: null,
      minDate: null,
      maxDate: null
    };
  }

  componentDidUpdate() {
    const {popupOpen} = this.state;

    if (popupOpen) {
      document.addEventListener('click', this.hidePopup);
      document.addEventListener('keydown', this.closeOnEscape);
    } else {
      document.removeEventListener('click', this.hidePopup);
      document.removeEventListener('keydown', this.closeOnEscape);
    }
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.hidePopup);
    document.removeEventListener('keydown', this.closeOnEscape);
  }

  render() {
    return <center className="DateFields">
      <div className='DateFields__inputContainer'>
        <DateInput
            className={'DateInput__start' + (this.isFieldSelected('startDate') ? ' DateInput__start--highlight' : '')}
            format={this.props.format}
            onDateChange={this.setStartDate}
            onClick={this.toggleDateRangeForStart}
            date={this.props.startDate} />
        <span>to</span>
        <DateInput
            className={'DateInput__end' + (this.isFieldSelected('endDate') ? ' DateInput__start--highlight' : '')}
            reference={this.saveEndDateField}
            format={this.props.format}
            onDateChange={this.setEndDate}
            onClick={this.toggleDateRangeForEnd}
            date={this.props.endDate} />
      </div>
      {
        this.state.popupOpen && (
          <div onClick={this.stopClosingPopup} className={
            'DateFields__range' +
            (this.isFieldSelected('startDate') ? ' DateFields__range--left' : '') +
            (this.isFieldSelected('endDate') ? ' DateFields__range--right' : '')
          }>
            <DateRange
                format={this.props.format}
                onDateChange={this.onDateRangeChange}
                startDate={this.props.startDate}
                endDate={this.props.endDate} />
          </div>
        )
      }
    </center>;
  }

  closeOnEscape = event => {
    if (event.key === 'Escape') {
      this.hidePopup();
    }
  }

  saveEndDateField = input => this.endDateField = input;

  onDateRangeChange = date => {
    this.props.onDateChange(this.state.currentlySelectedField, date);

    if (this.isFieldSelected('startDate')) {
      this.setState({
        currentlySelectedField: 'endDate',
      });
      this.endDateField.focus();
    } else {
      setTimeout(this.hidePopup, 350);
    }
  }

  stopClosingPopup = ({nativeEvent: event}) =>  {
    // https://stackoverflow.com/questions/24415631/reactjs-syntheticevent-stoppropagation-only-works-with-react-events
    event.stopImmediatePropagation();
  }

  hidePopup = () => {
    this.setState({
      popupOpen: false,
      currentlySelectedField: null,
      minDate: null,
      maxDate: null
    });
  }

  isFieldSelected(field) {
    return this.state.currentlySelectedField === field;
  }

  setStartDate = date => this.props.onDateChange('startDate', date);
  setEndDate = date => this.props.onDateChange('endDate', date);

  toggleDateRangeForStart = event => this.toggleDateRangePopup(event, 'startDate');
  toggleDateRangeForEnd = event => this.toggleDateRangePopup(event, 'endDate');

  toggleDateRangePopup = (event, field) => {
    this.stopClosingPopup(event);

    if (this.state.popupOpen) {
      return this.setState({
        currentlySelectedField: field
      });
    }

    this.setState({
      popupOpen: true,
      currentlySelectedField: field
    });
  }
}
