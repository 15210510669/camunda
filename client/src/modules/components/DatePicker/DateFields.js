/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';
import classnames from 'classnames';
import {parseISO} from 'date-fns';

import {format} from 'dates';

import DateRange from './DateRange';
import PickerDateInput from './PickerDateInput';
import {isDateValid} from './service';

import './DateFields.scss';

const POPPUP_CLASSNAME = 'dateRangeContainer';

export default class DateFields extends React.PureComponent {
  state = {
    popupOpen: false,
    currentlySelectedField: null,
  };

  dateFields = React.createRef();

  endDateField = React.createRef();

  // Modals stop propagation of events on elements outside the modal
  // Therefore, we need to attach the events on the modal instead of document
  getContext = () => this.dateFields.current?.closest('.Modal') || document;

  componentDidUpdate() {
    const {popupOpen} = this.state;

    const context = this.getContext();

    if (popupOpen) {
      context.addEventListener('click', this.hidePopup);
      context.addEventListener('keydown', this.closeOnEscape);
    } else {
      context.removeEventListener('click', this.hidePopup);
      context.removeEventListener('keydown', this.closeOnEscape);
    }
  }

  componentWillUnmount() {
    const context = this.getContext();
    context.removeEventListener('click', this.hidePopup);
    context.removeEventListener('keydown', this.closeOnEscape);
  }

  handleKeyPress = (evt) => {
    if (this.state.popupOpen && evt.key === 'Escape') {
      evt.stopPropagation();
    }
  };

  render() {
    const {startDate, endDate, forceOpen, type} = this.props;

    const startDateObj = parseISO(startDate);
    const endDateObj = parseISO(endDate);

    return (
      <div className="DateFields" ref={this.dateFields} onKeyDown={this.handleKeyPress}>
        <div className="inputContainer">
          {type !== 'before' && (
            <PickerDateInput
              className={classnames({
                highlight: this.isFieldSelected('startDate'),
              })}
              onChange={this.setDate('startDate')}
              onFocus={() => {
                this.setState({currentlySelectedField: 'startDate'});
              }}
              onSubmit={this.submitStart}
              onClick={() => this.toggleDateRangePopup('startDate')}
              value={startDate}
              isInvalid={!isDateValid(startDate)}
            />
          )}
          {type !== 'after' && (
            <PickerDateInput
              className={classnames({
                highlight: this.isFieldSelected('endDate'),
              })}
              ref={this.endDateField}
              onChange={this.setDate('endDate')}
              onFocus={() => {
                this.setState({currentlySelectedField: 'endDate'});
              }}
              onSubmit={this.submitEnd}
              onClick={() => this.toggleDateRangePopup('endDate')}
              value={endDate}
              isInvalid={!isDateValid(endDate)}
            />
          )}
        </div>
        {(this.state.popupOpen || forceOpen) && (
          <div
            className={classnames(POPPUP_CLASSNAME, {
              dateRangeContainerLeft: this.isFieldSelected('startDate'),
              dateRangeContainerRight: this.isFieldSelected('endDate'),
            })}
          >
            <DateRange
              type={type}
              endDateSelected={this.isFieldSelected('endDate')}
              onDateChange={this.onDateRangeChange}
              startDate={startDateObj}
              endDate={endDateObj}
            />
          </div>
        )}
      </div>
    );
  }

  submitStart = () => {
    if (this.props.type === 'between') {
      this.setState({currentlySelectedField: 'endDate'});
      this.endDateField.current.focus();
    } else {
      this.hidePopup();
    }
  };

  submitEnd = () => {
    this.hidePopup();
    this.endDateField.current.blur();
  };

  closeOnEscape = (event) => {
    if (event.key === 'Escape') {
      this.hidePopup();
    }
  };

  formatDate = (date) => (date ? format(date, this.props.format) : '');

  onDateRangeChange = ({startDate, endDate}) => {
    this.props.onDateChange('startDate', this.formatDate(startDate));
    this.props.onDateChange('endDate', this.formatDate(endDate));

    if (this.isFieldSelected('endDate') || this.props.type !== 'between') {
      setTimeout(this.hidePopup, 350);
    } else {
      this.setState({currentlySelectedField: 'endDate'});
      this.endDateField.current.focus();
    }
  };

  hidePopup = (evt) => {
    if (!evt?.target?.closest('.' + POPPUP_CLASSNAME)) {
      this.setState({
        popupOpen: false,
        currentlySelectedField: null,
      });
    }
  };

  isFieldSelected(field) {
    return this.state.currentlySelectedField === field;
  }

  setDate = (name) => (date) => this.props.onDateChange(name, date);

  toggleDateRangePopup = (field) => {
    this.setState({
      popupOpen: true,
      currentlySelectedField: field,
    });
  };
}
