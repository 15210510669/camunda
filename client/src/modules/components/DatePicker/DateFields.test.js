import React from 'react';
import moment from 'moment';

import DateFields from './DateFields';
import {mount} from 'enzyme';

jest.mock('./DateRange', () => props => `DateRange: props: ${JSON.stringify(props)}`);
jest.mock('./DateInput');

const format = 'YYYY-MM-DD';
const startDate = moment('2017-08-29').format(format);
const endDate = moment('2020-06-05').format(format);

it('should have start date input field', () => {
  const node = mount(<DateFields format={format} startDate={startDate} endDate={endDate} />);

  expect(node).toIncludeText('DateInput__start');
});

it('should have end date input field', () => {
  const node = mount(<DateFields format={format} startDate={startDate} endDate={endDate} />);

  expect(node).toIncludeText('DateInput__end');
});

it('should set startDate on date change of start date input field', () => {
  const spy = jest.fn();
  const node = mount(
    <DateFields format={format} startDate={startDate} endDate={endDate} onDateChange={spy} />
  );

  node.instance().setDate('startDate')('change');

  expect(spy).toBeCalledWith('startDate', 'change');
});

it('should set endDate on date change of end date input field', () => {
  const spy = jest.fn();
  const node = mount(
    <DateFields format={format} startDate={startDate} endDate={endDate} onDateChange={spy} />
  );

  node.instance().setDate('endDate')('change');

  expect(spy).toBeCalledWith('endDate', 'change');
});

it('should select date range popup on date input click', () => {
  const node = mount(
    <DateFields
      format={format}
      startDate={startDate}
      endDate={endDate}
      enableAddButton={jest.fn()}
    />
  );

  const evt = {nativeEvent: {stopImmediatePropagation: jest.fn()}};
  node.instance().toggleDateRangeForStart(evt);

  expect(evt.nativeEvent.stopImmediatePropagation).toHaveBeenCalled();
  expect(node.state('popupOpen')).toBe(true);
  expect(node.state('currentlySelectedField')).toBe('startDate');
});

it('should have DateRange', () => {
  const node = mount(
    <DateFields
      format={format}
      startDate={startDate}
      endDate={endDate}
      enableAddButton={jest.fn()}
    />
  );
  node.setState({popupOpen: true});

  expect(node).toIncludeText('DateRange');
});

it('should change currently selected date to endDate', () => {
  const spy = jest.fn();
  const node = mount(
    <DateFields
      format={format}
      startDate={startDate}
      endDate={endDate}
      onDateChange={spy}
      enableAddButton={jest.fn()}
    />
  );
  node.setState({popupOpen: true, currentlySelectedField: 'startDate'});

  node.instance().endDateField = document.createElement('input');
  node.instance().onDateRangeChange(moment(['2017-08-29']));

  expect(node.state('currentlySelectedField')).toBe('endDate');
});

it('should selected endDate after second selection', () => {
  const spy = jest.fn();
  const node = mount(
    <DateFields
      format={format}
      startDate={startDate}
      endDate={endDate}
      onDateChange={spy}
      enableAddButton={jest.fn()}
    />
  );
  node.setState({popupOpen: true, currentlySelectedField: 'startDate'});

  node.instance().endDateField = document.createElement('input');
  node.instance().onDateRangeChange(moment(['2017-08-29']));
  node.instance().onDateRangeChange(moment('2020-06-05'));

  expect(spy).toBeCalledWith('endDate', '2020-06-05');
});

it('should check if a start date or end date is valid', () => {
  const spy = jest.fn();
  const node = mount(
    <DateFields
      format={format}
      startDate={startDate}
      endDate={endDate}
      onDateChange={spy}
      enableAddButton={jest.fn()}
    />
  );
  const isValid = node.instance().isValid('startDate', moment('2017-08-29', format));
  const isNotValid = node.instance().isValid('endDate', moment('2017-06-05', format));
  expect(isValid).toBe(true);
  expect(isNotValid).toBe(false);
});
