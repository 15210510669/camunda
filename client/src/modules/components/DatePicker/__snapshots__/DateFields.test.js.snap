// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  className="DateFields"
  onKeyDown={[Function]}
>
  <div
    className="inputContainer"
  >
    <DateInput
      className=""
      isInvalid={false}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onSubmit={[Function]}
      value="2017-08-29"
    />
    <DateInput
      className=""
      isInvalid={false}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onSubmit={[Function]}
      value="2020-06-05"
    />
  </div>
</div>
`;
